Allow simple calculations within expand()

duncs opened the pull request:
Add in 'eval { ... }' within expand() to allow for simple calculations of
variables performed within a Safe compartment

    my $result = $cp->expand( 'eval{ ${V5} * {$V9} }');
    my $result = $cp->expand( 'eval { sprintf("%.2f", ${V12} / {$V5} ) }');

duncs committed:
Cater for parsing multiline traps

Some traps might have one element spread over multiple lines, such as:

hostname.somedomain
UDP: [192.168.100.10]:61613->[192.168.100.10]:162
DISMAN-EVENT-MIB::sysUpTimeInstance 0:0:00:00.01
SNMPv2-MIB::snmpTrapOID.0 saatrap::snmpSAAEvent
saatrap::saaEventName "Message Sent"
saatrap::saaEventDescription "Message Partner DataStoreIn, Session 2129 - Message sent
    Sequence number : 1
    UUMID           : ABCDABCDABCDABCDABCDABCD123
    Suffix          : 1059726480192
"
SNMP-COMMUNITY-MIB::snmpTrapAddress.0 192.168.100.10

This patch caters for reconstituting the lines where those lines are enclosed
with double quotes.  If only one quote is found from that point to the end of
the trap then the original packet is used, as now.

This patch also has the effect of fixing variable numbers as the multiple
lines are counted as variables in their own right, e.g. in the above $P11
is SNMP-COMMUNITY-MIB::snmpTrapAddress, not the expected $P7

duncs committed:
Fix double quoting on expanded trap variables

Give a line in the trap such as
  snmptrap::identifier "value"
and given a variable to expand such as
  my $var = $trap->expand('"${V3}"');
you get the following:
  $var = '""value""'
which $trap->eval will then fall over with

Look for double quoting (using double quotes only) in these situations and
remove one level of quote chars

duncs committed:
Revert "Fix double quoting on expanded trap variables"

This reverts commit f6baf7114f2b83921473b485501708e3a4239ad3.

duncs committed:
Allow simple calculations within expand()

Add in 'eval { ... }' within expand() to allow for simple calculations of
variables performed within a Safe compartment

    my $result = $cp->expand( 'eval{ ${V5} * {$V9} }');
    my $result = $cp->expand( 'eval { sprintf("%.2f", ${V12} / {$V5} ) }');

tonvoon commented:
The functionality here looks fine. Two things:
1) a value of 0 is being returned as an empty string. Test:

```
cmp_ok( $trap->expand('eval { ${V5} / 24 - 1 }'), 'eq', 0, "check for 0 value" );
```

2) Can you add a documentation section in the perldoc

duncs committed:
Improve tests

- Do a full trap comparison for the multi-line test
- Swap to 'done_testing();' to ignreo test count

duncs committed:
Merge remote-tracking branch 'upstream/master' into multiline_traps

duncs committed:
Merge remote-tracking branch 'origin/multiline_traps'

duncs committed:
Docs for eval within expand

duncs commented:
This pull request is based on the multiline traps in #1, so that will need merging first.

duncs committed:
Add in multi-line documentation

Cover values containing multiple lines and ensure double-quoting
is specified.
